src/mame/roland/roland_proe.cpp

```diff
@@ -0,0 +1,366 @@
+	m_ram_bank->configure_entries(0, 4, &m_ram[0], 0x2000);
+}
+
+void roland_proe_state::machine_reset()
+{
+	m_key_scan_i = 0;
+}
+
+void roland_proe_state::write_acia_clock(int state)
+{
+	m_acia->write_txc(state);
+	m_acia->write_rxc(state);
+}
+
+void roland_proe_state::int_midi_w(u8 data)
+{
+	// logerror("midi tx %02x\n", data);
+}
+
+void roland_proe_state::lcd_palette(palette_device &palette) const
+{
+	palette.set_pen_color(0, rgb_t(0, 0, 0));
+	palette.set_pen_color(1, rgb_t(0, 255, 0));
+}
+
+HD44780_PIXEL_UPDATE(roland_proe_state::pixel_update)
+{
+	if (x < 5 && y < 8 && line < 2 && pos < 16)
+		bitmap.pix(line * 8 + y, pos * 6 + x) = state;
+}
+
+void roland_proe_state::lcd_data_w(u8 data)
+{
+	if (m_lcd_mode == 0x04)
+		m_lcd->data_w(data);
+	else if (m_lcd_mode == 0xf8)
+		m_lcd->control_w(data);
+}
+
+void roland_proe_state::lcd_control_w(u8 data)
+{
+	// 0x04: data
+	// 0x06: ??
+	// 0xf8: control
+	if (data == 0x04 || data == 0x06 || data == 0xf8)
+		m_lcd_mode = data;
+}
+
+u8 roland_proe_state::lcd_r()
+{
+	return 0x00;
+}
+
+u8 roland_proe_state::keys_r()
+{
+	u8 result = m_keys[m_key_scan_i]->read() & 0xff;
+	return result;
+}
+
+u8 roland_proe_state::pa_r()
+{
+	u8 result = (m_keys[m_key_scan_i]->read() >> 8) | 0b11111000;
+	return result;
+}
+
+void roland_proe_state::pa_w(u8 data)
+{
+	m_key_scan_i = (data >> 3) & 0x07;
+}
+
+void roland_proe_state::leds_w(u8 data)
+{
+	// TODO
+	// logerror("leds %02x\n", data);
+}
+
+void roland_proe_state::bank_w(u8 data)
+{
+	m_rom_bank->set_entry(data >> 4);
+	m_ram_bank->set_entry(data & 3);
+}
+
+// u8 roland_proe_state::patterns_r(offs_t offset)
+// {
+// 	return m_patterns[(offset & 0x7fff) | (rom_bank << 15)];
+// }
+
+void roland_proe_state::proe_map_a(address_map &map)
+{
+	// 0x0000-0x00ff: cpu internal ram
+	// 0x1000-0x103f: cpu registers
+	map(0x1200, 0x1200).r(FUNC(roland_proe_state::keys_r));
+	map(0x1400, 0x1401).rw(m_acia, FUNC(acia6850_device::read), FUNC(acia6850_device::write));
+	map(0x160d, 0x160d).w(FUNC(roland_proe_state::lcd_control_w));
+	map(0x160e, 0x160e).rw(FUNC(roland_proe_state::lcd_r), FUNC(roland_proe_state::lcd_data_w));
+	map(0x180d, 0x180d).w(FUNC(roland_proe_state::bank_w));
+	map(0x180e, 0x180e).w(FUNC(roland_proe_state::leds_w));
+	map(0x1a00, 0x1bff).rw(m_keyscan, FUNC(mb63h149_device::read), FUNC(mb63h149_device::write));
+	map(0x2000, 0x3fff).bankrw("ram_bank");
+	map(0x4000, 0xbfff).bankrw("rom_bank");
+	map(0xc000, 0xffff).rom().region("cpu", 0); // ic4
+}
+
+void roland_proe_state::proe(machine_config &config)
+{
+	XTAL main_clock = 16_MHz_XTAL;
+
+	MC68HC11A0(config, m_cpu, main_clock / 2); // MC68HC11A0P
+	m_cpu->set_addrmap(AS_PROGRAM, &roland_proe_state::proe_map_a);
+	m_cpu->in_pa_callback().set(FUNC(roland_proe_state::pa_r));
+	m_cpu->out_pa_callback().set(FUNC(roland_proe_state::pa_w));
+	m_cpu->serial_tx_cb().set(FUNC(roland_proe_state::int_midi_w));
```

Comment:
We really, really don't want to hint that serial-faked-as-parallel is acceptable in new code.  Since you don't actually do anything in int_midi_w, please remove the callback until real serial is implemented.